In [1]:
import numpy as np 
import pandas as pd 

I choose first Sentiment Analysic in which will be analyzing the review to determine if the emotional tone of the message is positive, negative, or neutral. However, before that I would like to apply a normal structure for a NLP processing steps wise.


**Step 1: Read the Text Data**

In [2]:
import re
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords

In [3]:
import pandas as pd
import chardet

dataset_path = r"D:\Huyen\Data Science\Kaggle Projects\New_Delhi_reviews.csv"

# Read the file in binary mode and detect the encoding
with open(dataset_path, 'rb') as f:
    result = chardet.detect(f.read())

# Get the detected encoding
encoding = result['encoding']

# Read the file using the detected encoding
df = pd.read_csv(dataset_path, encoding=encoding)



In [4]:
df.head()

rating_review                                        review_full
0              5  Totally in love with the Auro of the place, re...
1              5  I went this bar 8 days regularly with my husba...
2              5  We were few friends and was a birthday celebra...
3              5  Fatjar Cafe and Market is the perfect place fo...
4              5  Hey Guys, if you are craving for pizza and sea...

In [5]:
df.shape

(999, 2)

In [6]:
#what portion of our review are distributed in the dataset

df['rating_review'].value_counts()

5    697
4    244
3     36
2     12
1     10
Name: rating_review, dtype: int64

In [7]:
#Check for missing value

print('Number of nulls in rating_review:{}'.format(df['rating_review'].isnull().sum()))
print('Number of nulls in review:{}'.format(df['review_full'].isnull().sum()))

Number of nulls in rating_review:0
Number of nulls in review:0


Above dataset has 2 null value in the review, consider that we have a good number of review, we can choose to remove this null value columns

In [8]:
df = df.dropna(subset =['review_full'])

In [9]:
df.shape

(999, 2)

**Pre-Processing Text Data**

In [10]:
#Remove Punctuation

import string

string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [11]:
def remove_punct(review_full):
    review_full ="".join([char for char in review_full if char not in string.punctuation])
    return review_full
df['review'] = df['review_full'].apply(lambda x: remove_punct(x))

df.head()

rating_review                                        review_full  \
0              5  Totally in love with the Auro of the place, re...   
1              5  I went this bar 8 days regularly with my husba...   
2              5  We were few friends and was a birthday celebra...   
3              5  Fatjar Cafe and Market is the perfect place fo...   
4              5  Hey Guys, if you are craving for pizza and sea...   

                                              review  
0  Totally in love with the Auro of the place rea...  
1  I went this bar 8 days regularly with my husba...  
2  We were few friends and was a birthday celebra...  
3  Fatjar Cafe and Market is the perfect place fo...  
4  Hey Guys if you are craving for pizza and sear...

In [12]:
#Tokenize

#Define a function to split our sentences into a list of words

import re

def tokenize (review_full):
    tokens = re.split('\W+', review_full)
    return tokens
df['review_tokenized'] = df['review'].apply(lambda x: tokenize(x.lower()))

df.head()

rating_review                                        review_full  \
0              5  Totally in love with the Auro of the place, re...   
1              5  I went this bar 8 days regularly with my husba...   
2              5  We were few friends and was a birthday celebra...   
3              5  Fatjar Cafe and Market is the perfect place fo...   
4              5  Hey Guys, if you are craving for pizza and sea...   

                                              review  \
0  Totally in love with the Auro of the place rea...   
1  I went this bar 8 days regularly with my husba...   
2  We were few friends and was a birthday celebra...   
3  Fatjar Cafe and Market is the perfect place fo...   
4  Hey Guys if you are craving for pizza and sear...   

                                    review_tokenized  
0  [totally, in, love, with, the, auro, of, the, ...  
1  [i, went, this, bar, 8, days, regularly, with,...  
2  [we, were, few, friends, and, was, a, birthday...  
3  [fatjar, cafe, and, market, is, the, perfect, ...  
4  [hey, guys, if, you, are, craving, for, pizza,...

In [13]:
#Remove Stopwords

import nltk
nltk.download('stopwords')


[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\HP\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [14]:
import nltk

stopwords = nltk.corpus.stopwords.words('english')


In [15]:
#Define a function to remove all stopwords

def remove_stopwords(tokenized_review_full):
    review_full = [word for word in tokenized_review_full if word not in stopwords]
    return review_full

df['review_nonstop'] = df['review_tokenized'].apply(lambda x: remove_stopwords(x))

df.head()

rating_review                                        review_full  \
0              5  Totally in love with the Auro of the place, re...   
1              5  I went this bar 8 days regularly with my husba...   
2              5  We were few friends and was a birthday celebra...   
3              5  Fatjar Cafe and Market is the perfect place fo...   
4              5  Hey Guys, if you are craving for pizza and sea...   

                                              review  \
0  Totally in love with the Auro of the place rea...   
1  I went this bar 8 days regularly with my husba...   
2  We were few friends and was a birthday celebra...   
3  Fatjar Cafe and Market is the perfect place fo...   
4  Hey Guys if you are craving for pizza and sear...   

                                    review_tokenized  \
0  [totally, in, love, with, the, auro, of, the, ...   
1  [i, went, this, bar, 8, days, regularly, with,...   
2  [we, were, few, friends, and, was, a, birthday...   
3  [fatjar, cafe, and, market, is, the, perfect, ...   
4  [hey, guys, if, you, are, craving, for, pizza,...   

                                      review_nonstop  
0  [totally, love, auro, place, really, beautiful...  
1  [went, bar, 8, days, regularly, husband, fully...  
2  [friends, birthday, celebration, food, good, t...  
3  [fatjar, cafe, market, perfect, place, casual,...  
4  [hey, guys, craving, pizza, searching, visit, ...

(Save for later use) Define a function to handle all data cleaning

def clean_text(text):
    text = " ".join([word.lower() for word in text if word not in string.punctuation])
    tokens = re.split('\W+', text)
    text = [wor for word in tokens if word not in stopwords]
    return text

In [16]:
#create function to clean text

def clean_text(review_full):
    review_full = "".join([word.lower() for word in review_full if word not in string.punctuation])
    tokens = re.split('\W+', review_full)
    review_full = [word for word in tokens if word not in stopwords]
    return review_full

In [19]:
#Fit a basic TFIDF Vertorize and view the results

from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vect = TfidfVectorizer(analyzer=clean_text)
X_tfidf = tfidf_vect.fit_transform(df['review_full'])

print(X_tfidf.shape)

feature_names = tfidf_vect.get_feature_names_out()
print(feature_names)



(999, 6510)
['' '01133036298' '1' ... 'zone' 'zucchini' 'éclair']


In [22]:
import string
import re
from sklearn.feature_extraction.text import TfidfVectorizer

def clean_text(review_full):
    review_full = "".join([word.lower() for word in review_full if word not in string.punctuation])
    tokens = re.split('\W+', review_full)
    stopwords = ['review_nonstop']  # Define your list of stopwords here
    review_full = [word for word in tokens if word not in stopwords]
    return review_full

tfidf_vect = TfidfVectorizer(analyzer=clean_text)
X_tfidf = tfidf_vect.fit_transform(df['review_full'])

print(X_tfidf.shape)

feature_names = tfidf_vect.get_feature_names_out()
print(feature_names)



(999, 6644)
['' '01133036298' '1' ... 'zone' 'zucchini' 'éclair']


In [23]:
X_tfidf

<999x6644 sparse matrix of type '<class 'numpy.float64'>'
	with 52665 stored elements in Compressed Sparse Row format>

In [24]:
X_features = pd.DataFrame (X_tfidf.toarray())

In [25]:
X_features.head()

0     1     2     3     4     5     6     7     8     9     ...  6634  \
0  0.000000   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
1  0.121570   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2  0.000000   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
3  0.075442   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
4  0.060349   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   

   6635  6636  6637  6638  6639  6640  6641  6642  6643  
0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[5 rows x 6644 columns]

**Explore RandomForestClassifier Attributes & Hyperparameters**

In [26]:
#import Random Foresh for Classification from sklearn

from sklearn.ensemble import RandomForestClassifier

rf_classifier = RandomForestClassifier(n_estimators=100, max_depth=10, random_state=42)
print(rf_classifier)

RandomForestClassifier(max_depth=10, random_state=42)


In [27]:
from sklearn.metrics import precision_score, recall_score
from sklearn.model_selection import train_test_split

In [28]:
X_train, X_test, y_train, y_test = train_test_split(X_features, df['rating_review'], test_size=0.2)

In [29]:
#fit a basic Random Forest Model

rf = RandomForestClassifier()
rf_model = rf.fit(X_train, y_train)

In [30]:
#make predictions on the test set using the fit model
y_pred = rf_model.predict(X_test)

In [31]:
#Evaluate model prediction using precision & recall

from sklearn.metrics import precision_score, recall_score

precision = precision_score(y_test, y_pred, pos_label='5', average='macro')
recall = recall_score(y_test, y_pred, pos_label='5', average='macro')

print('Precision: {}/ Recall: {}'.format(round(precision, 3), round(recall, 3)))


Precision: 0.259/ Recall: 0.208


C:\Users\HP\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1396: UserWarning: Note that pos_label (set to '5') is ignored when average != 'binary' (got 'macro'). You may use labels=[pos_label] to specify a single positive class.
  warnings.warn(
C:\Users\HP\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1396: UserWarning: Note that pos_label (set to '5') is ignored when average != 'binary' (got 'macro'). You may use labels=[pos_label] to specify a single positive class.
  warnings.warn(


In Summary of the above code: 

Precision measures the ability of a classifier to correctly identify positive samples among the samples it predicted as positive. A precision score of 0.259 indicates that out of the samples predicted as positive, approximately 25.9% were actually positive.

Recall, also known as sensitivity or true positive rate, measures the ability of a classifier to correctly identify positive samples among all the actual positive samples. A recall score of 0.208 suggests that approximately 20.8% of the actual positive samples were correctly identified by the classifier.

Both precision and recall scores range from 0 to 1, with higher values indicating better performance.

The following we will do the same for Rating 4

In [32]:
#Evaluate model prediction using precision & recall

from sklearn.metrics import precision_score, recall_score

precision = precision_score(y_test, y_pred, pos_label='4', average='macro')
recall = recall_score(y_test, y_pred, pos_label='4', average='macro')

print('Precision: {}/ Recall: {}'.format(round(precision, 3), round(recall, 3)))


Precision: 0.259/ Recall: 0.208


C:\Users\HP\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1396: UserWarning: Note that pos_label (set to '4') is ignored when average != 'binary' (got 'macro'). You may use labels=[pos_label] to specify a single positive class.
  warnings.warn(
C:\Users\HP\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\HP\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1396: UserWarning: Note that pos_label (set to '4') is ignored when average != 'binary' (got 'macro'). You may use labels=[pos_label] to specify a single positive class.
  warnings.warn(


 Predict review rating from dataset of Review on the website

In [33]:

from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score

X = df['review_full']  # Reviews as input features
y = df['rating_review']  # Review ratings as target variable

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 3: Preprocess the text using TF-IDF vectorization
vectorizer = TfidfVectorizer()
X_train_vectors = vectorizer.fit_transform(X_train)
X_test_vectors = vectorizer.transform(X_test)

# Step 4: Train a machine learning model
model = LinearSVC()
model.fit(X_train_vectors, y_train)

# Step 5: Make predictions on the test set
y_pred = model.predict(X_test_vectors)

# Step 6: Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)


Accuracy: 0.685


In [34]:
from sklearn.metrics import precision_score, recall_score, f1_score

# Calculate precision
precision = precision_score(y_test, y_pred, average='weighted')

# Calculate recall
recall = recall_score(y_test, y_pred, average='weighted')

# Calculate F1 score
f1 = f1_score(y_test, y_pred, average='weighted')

print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)


Precision: 0.6619163763066201
Recall: 0.685
F1 Score: 0.636491043203372


C:\Users\HP\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Precision: Precision measures the proportion of correctly predicted positive instances (true positives) out of all instances predicted as positive. In your case, a precision of 0.6619163763066201 means that around 66.2% of the reviews predicted as positive were actually positive.

Recall: Recall (also known as sensitivity) measures the proportion of correctly predicted positive instances (true positives) out of all actual positive instances. With a recall of 0.685, it means that approximately 68.5% of the actual positive reviews were correctly identified.

F1 Score: The F1 score combines precision and recall into a single metric, providing a balance between the two. It is the harmonic mean of precision and recall. In your case, an F1 score of 0.636491043203372 indicates the overall performance of the model, considering both precision and recall.

#### In short, we should not use this model for further prediction on the review need further tuning. As part of lesson we will not proceed this model.